### ISSUE TITLE
Support non-bmi2 cpus

### ISSUE THREAD

**Issue by clbr** [MartinezTorres/marlin#7]
The current code requires a BMI2 cpu (Haswell or Excavator).

**Comment by MartinezTorres** [comment]
This is correct, I was investigating fast ways to do the bit shuffling without BMI2 but I've haven't found a compact and portable code yet, and a naive version is just very slow. 

I'm open to suggestions, which platform do you aim at currently?

**Comment by clbr** [comment]
My dev box is a Phenom II, and I have POWER and Mips setups too. Anyway just wanted to raise the issue, I'm still investigating Marlin, whether it works for my uses. Slow encoding is irrelevant for my purposes, I'm after fast decoding while keeping Rice-level compression.

**Comment by clbr** [comment]
Oh, adding that the Mips are BE, others LE. I guess this hasn't been tested on big-endian at all yet.

**Comment by MartinezTorres** [comment]
The code is definitely not tested on big-endian architectures. While it is cleaner than the original code, it is still "research code", meant to investigate those dictionaries. I have less inclined to investigate BE architectures because, after ARM migrated basically to LE, BE remains a small minority.
I'll investigate non-BMI cpus, but I've tested on some first generation Ryzen and I found them quite slower than Intels, in general, at least when running Marlin.
I have no MIPS or Power available to test BE architectures.

**Comment by clbr** [comment]
It seems Marlin doesn't work for my case, so no need to spend time on the non-bmi2 path for me.

Details, if you're interested:
I tried replacing an existing Rice-type bit-run-length encoder with wholesale Marlin. First brute-forcing with all prebuilt dictionaries, the best such gave 2-7x bigger files than the baseline. Creating a custom dictionary for each 32kb data block, which is infeasible due to the dict size, resulted in ~2x bigger files than baseline.

So even in the best case, with a perfect dictionary, it couldn't match the baseline. Now my data is clearly not a perfect memoryless source, other entropy encoders may do badly as well. However adding a modeling layer would bring the decoding speed down vs plain Marlin.

For comparison, zlib level 9 created 5-10% bigger files than the baseline. Perhaps trying zlib-type LZ but with Huffman replaced by Marlin would be worth a try.

**Comment by MartinezTorres** [comment]
The issues you mention are on the pipeline for future investigations, seeing that there is interest in those, I'll prioritize them within my research pipeline.  Thanks for trying Marlin!
